File: framework/languages/java/e2e/RuntimeMonitoring/Server/Service/src/main/java/systems/zlink/e2e/runtimemonitoring/service/handlers/MonitoringEventHandlers.java
package systems.zlink.e2e.runtimemonitoring.service.handlers;

import systems.zlink.e2e.runtimemonitoring.service.support.EvidenceState;
import systems.zlink.framework.monitoring.ZLinkLocationRuntimeEvent;
import systems.zlink.framework.monitoring.ZLinkRuntimeEventHandler;
import systems.zlink.framework.monitoring.ZLinkSocketEvent;
import systems.zlink.framework.monitoring.ZLinkSpotEvent;

public final class MonitoringEventHandlers {
    private MonitoringEventHandlers() {
    }

    public static final class SocketRecorder implements ZLinkRuntimeEventHandler<ZLinkSocketEvent> {
        private final EvidenceState state;

        public SocketRecorder(EvidenceState state) {
            this.state = state;
        }

        @Override
        public void handle(ZLinkSocketEvent event) {
            state.record(
                "socket",
                event.sourceName(),
                event.event().name(),
                event.localAddr() + "|" + event.remoteAddr()
                    + "|" + event.routingId().map(Object::toString).orElse(""));
        }
    }

    public static final class FailingSocketRecorder
        implements ZLinkRuntimeEventHandler<ZLinkSocketEvent> {
        private final EvidenceState state;

        public FailingSocketRecorder(EvidenceState state) {
            this.state = state;
        }

        @Override
        public void handle(ZLinkSocketEvent event) {
            state.record(
                "monitoring",
                event.sourceName(),
                "HandlerFailureInjected",
                event.event().name());
            throw new IllegalStateException("intentional monitoring handler failure");
        }
    }

    public static final class LocationRuntimeRecorder
        implements ZLinkRuntimeEventHandler<ZLinkLocationRuntimeEvent> {
        private final EvidenceState state;

        public LocationRuntimeRecorder(EvidenceState state) {
            this.state = state;
        }

        @Override
        public void handle(ZLinkLocationRuntimeEvent event) {
            state.record(
                "location",
                event.sourceName(),
                event.event().name(),
                "topology=" + (event.topology() == null ? 0 : event.topology().size())
                    + "|summary=" + (event.serviceSummary() == null ? 0 : event.serviceSummary().size()));
        }
    }

    public static final class SpotRecorder implements ZLinkRuntimeEventHandler<ZLinkSpotEvent> {
        private final EvidenceState state;

        public SpotRecorder(EvidenceState state) {
            this.state = state;
        }

        @Override
        public void handle(ZLinkSpotEvent event) {
            state.record(
                "spot",
                event.sourceName(),
                event.event().name(),
                "peers=" + event.peers().size() + "|subjects=" + event.subjects().size()
                    + event.timerDiagnostic().map(value -> "|timer=" + value).orElse(""));
        }
    }
}

File: framework/languages/java/e2e-kotlin/RuntimeMonitoring/Server/Service/src/main/kotlin/systems/zlink/e2e/kotlin/runtimemonitoring/service/MonitoringEventHandlers.kt
package systems.zlink.e2e.kotlin.runtimemonitoring.service

import systems.zlink.framework.monitoring.ZLinkLocationRuntimeEvent
import systems.zlink.framework.monitoring.ZLinkRuntimeEventHandler
import systems.zlink.framework.monitoring.ZLinkSocketEvent
import systems.zlink.framework.monitoring.ZLinkSpotEvent

class MonitoringEventHandlers private constructor() {
    class SocketRecorder(
        private val state: EvidenceState,
    ) : ZLinkRuntimeEventHandler<ZLinkSocketEvent> {
        override fun handle(event: ZLinkSocketEvent) {
            state.record(
                "socket",
                event.sourceName(),
                event.event().name,
                "${event.localAddr()}|${event.remoteAddr()}|${event.routingId().map { it.toString() }.orElse("")}",
            )
        }
    }

    class FailingSocketRecorder(
        private val state: EvidenceState,
    ) : ZLinkRuntimeEventHandler<ZLinkSocketEvent> {
        override fun handle(event: ZLinkSocketEvent) {
            state.record(
                "monitoring",
                event.sourceName(),
                "HandlerFailureInjected",
                event.event().name,
            )
            throw IllegalStateException("intentional monitoring handler failure")
        }
    }

    class LocationRuntimeRecorder(
        private val state: EvidenceState,
    ) : ZLinkRuntimeEventHandler<ZLinkLocationRuntimeEvent> {
        override fun handle(event: ZLinkLocationRuntimeEvent) {
            state.record(
                "location",
                event.sourceName(),
                event.event().name,
                "topology=${event.topology()?.size ?: 0}|summary=${event.serviceSummary()?.size ?: 0}",
            )
        }
    }

    class SpotRecorder(
        private val state: EvidenceState,
    ) : ZLinkRuntimeEventHandler<ZLinkSpotEvent> {
        override fun handle(event: ZLinkSpotEvent) {
            val timerDetail = event.timerDiagnostic().map { "|timer=$it" }.orElse("")
            state.record(
                "spot",
                event.sourceName(),
                event.event().name,
                "peers=${event.peers().size}|subjects=${event.subjects().size}$timerDetail",
            )
        }
    }
}
